server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistEventsStep.java

```diff
@@ -31,111 +41,171 @@
 import org.sonar.server.computation.component.Component;
 import org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor;
 import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.event.Event;
+import org.sonar.server.computation.event.EventRepository;
 import org.sonar.server.db.DbClient;
 
-import static org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor.Order.PRE_ORDER;
+import static com.google.common.collect.Iterables.transform;
 
 public class PersistEventsStep implements ComputationStep {
 
   private final DbClient dbClient;
   private final System2 system2;
-  private final TreeRootHolder treeRootHolder;
   private final BatchReportReader reportReader;
+  private final TreeRootHolder treeRootHolder;
+  private final EventRepository eventRepository;
 
-  public PersistEventsStep(DbClient dbClient, System2 system2, TreeRootHolder treeRootHolder, BatchReportReader reportReader) {
+  public PersistEventsStep(DbClient dbClient, System2 system2, TreeRootHolder treeRootHolder, BatchReportReader reportReader,
+    EventRepository eventRepository) {
     this.dbClient = dbClient;
     this.system2 = system2;
     this.treeRootHolder = treeRootHolder;
     this.reportReader = reportReader;
+    this.eventRepository = eventRepository;
   }
 
   @Override
   public void execute() {
-    DbSession session = dbClient.openSession(false);
+    final DbSession session = dbClient.openSession(false);
     try {
-      new EventVisitor(session, reportReader.readMetadata().getAnalysisDate()).visit(treeRootHolder.getRoot());
+      final long analysisDate = reportReader.readMetadata().getAnalysisDate();
+      new DepthTraversalTypeAwareVisitor(Component.Type.FILE, DepthTraversalTypeAwareVisitor.Order.PRE_ORDER) {
+        @Override
+        public void visitProject(Component project) {
+          processComponent(project, session, analysisDate);
+        }
+
+        @Override
+        public void visitModule(Component module) {
+          processComponent(module, session, analysisDate);
+        }
+
+        @Override
+        public void visitDirectory(Component directory) {
+          processComponent(directory, session, analysisDate);
+        }
+
+        @Override
+        public void visitFile(Component file) {
+          processComponent(file, session, analysisDate);
```

Comment:
well, we could make that optimization, it would work since, for now, we only create events on projects (not even sure on modules).

But design allows defining events on any component, if we want this step to comply with the CE design, we should saves events on any Component. The extra cost can be estimated as little so I'd rather save on any Component until proven otherwise.